### PR TITLE
Remove stripe decorator

### LINKED ISSUE #7326: Reduce/remove ActiveMerchant Stripe decorator
### What we should change and why (this is tech debt)

We customize some code from ActiveMerchant by using a decorator; ideally we would reduce how much we have to customize and just use the standard gem behavior. 

### Context

https://github.com/openfoodfoundation/openfoodnetwork/pull/7286

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- Closes #7326
- Superseed: #11880
- Partially address: #11670

This is based on the work done in #11880

As explained in #7326, we customised some code from ActiveMerchant, this is not needed any more. It will make further upgrade of ActiveMerchant easier. 

I also refactored some model payment spec that somehow were depending on this. I remove the dependency and any stubbing of Stripe request.

#### What should we test?

:warning: **This are the scenario I am aware of that are affected by this change, please test anything that I might have missed.**

First make sure you have a shop with stripe connect setup. In the back office,  check Configuration -> Payment methods
It looks like this when the payment method is correctly configured :
Screenshot Capture - 2024-02-13 - 13-41-12

##### capture payment
- place an order using stripe connect ( you can use any of these testing card https://stripe.com/docs/testing?testing-method=card-numbers) 
   --> placing an order is successful
- check in the back office that the order has a payment state of "PAID ($0.00)"

##### refund payment
- place an order using stripe connect ( you can use any of these testing card https://stripe.com/docs/testing?testing-method=card-numbers)
- In the back office, go to the order payment page, easiest way to do that is to click on the payment step from the order list  
- Click on the X button
- The payment status should become "Void"

##### store a card
- place an order using stripe connect ( you can use any of these testing card https://stripe.com/docs/testing?testing-method=card-numbers), and tick "Save card for future use"
 - Go to the customer account page and click on "Credit Card" tab
   --> check there is a saved card matching the one you just used
 - Click on the delete button
   --> Card is successfully deleted 

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: spec/models/spree/gateway/stripe_sca_spec.rb
Comment (rioug) [inline spec/models/spree/gateway/stripe_sca_spec.rb:80]: @filipefurtad0 I am trying to test Stripe refund here but I am a bit blocked 
The `void` method retrieve the payment intent here : https://github.com/openfoodfoundation/openfoodnetwork/blob/4dda4e2385fcf45b95f4c2bfe745b2a98b7d42af/app/models/spree/gateway/stripe_sca.rb#L90-L91
It requires a `stripe_account_id`. The problem here is I don't know which `stripe_account_id` to use, I am using the Stripe test account you setup. Currently I get the following error :
```
 1) Spree::Gateway::StripeSCA#void with a confirmed payment refunds the payment
     Failure/Error:
       payment_intent_response = Stripe::PaymentIntent.retrieve(payment_intent_id,
                                                                stripe_account: stripe_account_id)
     
     Stripe::PermissionError:
       The provided key 'sk_test_******************************uCJm' does not have access to account 'acc_123' (or that account does not exist). Application access may have been revoked.
```
I feel like we might need to create a payment intent linked to a test customer, maybe adding `customer:` here (see [doc](https://stripe.com/docs/api/payment_intents/create)) : 
https://github.com/openfoodfoundation/openfoodnetwork/blob/4dda4e2385fcf45b95f4c2bfe745b2a98b7d42af/spec/models/spree/gateway/stripe_sca_spec.rb#L42-L49

What do you think ? Do you have a test customer I could use on your stripe test account ?

Comment (filipefurtad0) [reply spec/models/spree/gateway/stripe_sca_spec.rb:80]: Thanks for pinging me here @rioug . I think:
- we need to create customers by demand (with real Stripe calls), so that we don't use hard coded customer IDs.
- the error message points to `The provided key 'sk_test_******************************uCJm' does not have access to account 'acc_123'` because we're setting it with this value a few lines above. I think we need to create connected accounts as well, when needed (as above, with real Stripe calls), and then use the ID on line 90 above

I've spotted an inconsistency in a related spec, and try to fix it [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/12127/files#). I think the setup could be used in this spec as well.

I hope this helps. What do you think?

Comment (rioug) [reply spec/models/spree/gateway/stripe_sca_spec.rb:80]: Thanks @filipefurtad0 I made progress on this but not quite there yet, and as David pointed out I can use my own stripe test account to get the test working, it should unblock me.

**Review by mkllnk — APPROVED** [review]
Excellent, great commit history. Thank you.

I think that we need to re-record the cassettes though as this was merged:

* #12141

**Comment by rioug** [comment]
> I think that we need to re-record the cassettes though as this was merged:
> 
>     * [chore(deps): bump stripe from 10.6.0 to 10.8.0 #12141](https://github.com/openfoodfoundation/openfoodnetwork/pull/12141)

Indeed, and I seem to have committed some 10.2.0 cassettes as well, or most likely forgot to remove them after the rebase

File: spec/system/consumer/checkout/payment_spec.rb
Comment (filipefurtad0) [inline spec/system/consumer/checkout/payment_spec.rb:257]: Hum, interesting! You're using Stripe stubs on a system spec :eyes: 
I think I've seen this before, but I have not gotten into it yet.

I wonder if instead of these stubs we could not use VCR (as for example, on [this spec](https://github.com/openfoodfoundation/openfoodnetwork/commit/baef7de663bcc2d657d9d7063186b3dbe3566a8a))

In any case, not a blocker - just wondering.

**Review by filipefurtad0 — APPROVED** [review]
Thanks so much @rioug :muscle: 
I've left some minor comments on the specs, we can address these later on. if needed.

Approving, but pinging @dacook for code review for the changes on the decorator itself.

I'd suggest to mark this PR as user facing, because it partially/possibly addresses #11670.

Also, I think we should stage and test this one manually, to:
- verify the [previously failing scenario](https://github.com/openfoodfoundation/openfoodnetwork/pull/11880#issuecomment-1834335037)
- re-test https://github.com/openfoodfoundation/openfoodnetwork/issues/11670 and see if we can close it

File: spec/models/spree/gateway/stripe_sca_spec.rb
Comment (dacook) [inline spec/models/spree/gateway/stripe_sca_spec.rb:128]: This setup is almost the same as the the above, so it would be great to DRY it a bit. If `payment` and `payment_intent` are moved to `let`s, the `payment` can be shared at a higher level.

File: spec/models/spree/payment_spec.rb
Comment (dacook) [inline spec/models/spree/payment_spec.rb:498]: These two specs are in the same `context`, but seem to duplicate the setup. can it be moved to `let`/`before` blocks?
The second example doesn't seem to need as much setup as the first, so I wonder if it's all necessary.

**Review by dacook — APPROVED** [review]
Great, I won't profess to have understood _all_ of it, but the changes look appropriate to me.

I have some comments on further tidying specs, but they're not essential. 

Filipe's comment about not using a hardcoded account ID sounds like a good idea. I don't follow it exactly so will leave it up to you. Moving to in dev just in case you want to address any of these.

File: spec/system/consumer/checkout/payment_spec.rb
Comment (rioug) [reply spec/system/consumer/checkout/payment_spec.rb:257]: I should have added a comment, from the commit message : 
> I used stripe stubs for the new scenario because storing a card on
> stripe depends on some client side interaction with Stripe. We can't
> capture that with VCR.

That's my understanding at least, I might have missed something

File: spec/models/spree/gateway/stripe_sca_spec.rb
Comment (rioug) [reply spec/models/spree/gateway/stripe_sca_spec.rb]: Good point :+1:  , fixed now.

**Comment by rioug** [comment]
> I'd suggest to mark this PR as user facing, because it partially/possibly addresses #11670.
> 
> Also, I think we should stage and test this one manually, to:
>  * verify the [previously failing scenario](https://github.com/openfoodfoundation/openfoodnetwork/pull/11880#issuecomment-1834335037)

This covered in the "Store a card" scenario described above.
>  * re-test [[Stripe, BO] Can't fully refund Stripe payments on complete orders #11670](https://github.com/openfoodfoundation/openfoodnetwork/issues/11670) and see if we can close it

I am working on a PR for this one, this PR fix the underlying problem, ie the refund failing. But it doesn't fix the fact that the error is not shown to the user.

**Review by mkllnk — APPROVED** [review]
I think you addressed all the comments. Moving to test-ready.

**Comment by filipefurtad0** [comment]
Hey @rioug ,

Thanks a lot for all the detailed notes on what to test. Followed the proposed steps below:

1. **capture payment**

- place an order using stripe connect ( you can use any of these testing card https://stripe.com/docs/testing?testing-method=card-numbers)
- placing an order is successful - :green_circle: 
- check in the back office that the order has a payment state of "PAID ($0.00)" - :green_circle: 


2.  **refund payment**

- place an order using stripe connect ( you can use any of these testing card https://stripe.com/docs/testing?testing-method=card-numbers)
- In the back office, go to the order payment page, easiest way to do that is to click on the payment step from the order list
- Click on the X button
- The payment status should become "Void"  - :green_circle:  

Whoa. This now works:

image

This is confirmed by checking the respective account on Stripe:

image


3. **store a card**

- place an order using stripe connect ( you can use any of these testing card https://stripe.com/docs/testing?testing-method=card-numbers), and tick "Save card for future use"
- Go to the customer account page and click on "Credit Card" tab
    --> check there is a saved card matching the one you just used  - :green_circle:  
- Click on the delete button
    --> Card is successfully deleted  - :green_circle:  

Awesome :tada: 

Merging.

**Pull request merged**
